### ISSUE TITLE
Practical use of CudnnRNN

### ISSUE THREAD

**Issue by georgesterpu** [tensorflow/tensorflow#24469]
Hi @houtoms @protoget @ebrevdo 

I imagined that #23588 would replace TensorFlow's `LSTMCell` implementation, but some main limitations at the moment are the `RNNCell` wrappers such as Dropout, Attention or Residual, all frequently used in research projects and essential for learning. Even if cuDNN was open-source, it would still be impractical for people to code research ideas in c++ and compile the toolkit every time.

Would it be possible to make `LSTMCell` run as fast as `CudnnLSTM` while retaining the flexibility of the Python API ? Could XLA-generated code match the performance of cudnn given sufficient effort from both TF and Nvidia ? Or are there any plans to support these popular features in cuDNN soon ?

**Comment by protoget** [comment]
Thank you for the comment.

#23588 does not change `LSTMCell`, but adds support if using `tf.contrib.cudnn_rnn.LSTM`. There's no feature parity between TF LSTM and CudnnLSTM.

Users could use `UnifiedLSTM` in Keras (probably in the next release, we just implemented internally) which would pick CudnnLSTM if the LSTM config is satisifed in CudnnLSTM. You're right at present , feature-wise, CudnnLSTM is a subset of regular LSTM.

I guess your goal is to have a backend that can pattern-match LSTM in the graph and replace it with high-perf LSTM kernels. I'm sure XLA is interested. I'm not aware of NV's plan, regarding if 1) NV would OSS Cudnn 2) How NV's compile toolchain interacts with XLA. I'll add relevant ppl in the thread to comment.

@azaks2 
@jlebar

**Comment by jlebar** [comment]
Getting XLA:GPU to perform well on LSTMs is definitely on our roadmap for next year.  No promises that we'll be able to match the performance of cudnn, as nvidia gets to do some things (like access the machine code, SASS) that we can't.  Nonetheless we are working very closely with nvidia on XLA.

Fundamentally most of our work internally is driven by Google's needs.  That means I cannot commit to doing any particular optimization for any particular models within any particular timeframe.

If this is important to you and you're interested in contributing, we'd definitely welcome that and help you get started.

**Comment by jlebar** [comment]
@ymodak Based on my comment above, I'm confused what caused this to be assigned to me.  Would you be willing to elaborate on what further action you're looking for from me?

**Comment by ymodak** [comment]
@jlebar Currently I am waiting for the user to respond. If he's willing to make a contribution then it will require your inputs. At this point we don't need any action from your side. I will reassign it to you/other assignee depending on its progress. Thanks!

**Comment by jlebar** [comment]
@ymodak Since we are not looking for any action from me, would it be OK if the bug was not assigned to me?

This way I can maintain the invariant that when I look at the set of bugs assigned to me, it's the set of things that need action from me.  This is helpful for keeping things from falling through the cracks.

**Comment by georgesterpu** [comment]
Hi,
Thanks a lot for the explanations, and apologies if my lack of response led to a sub-optimal workflow.
I mainly wanted to see if the updated `CudnnRNN` API could be used in machine translation or speech recognitions applications, and, if not, which of its aspects could further improve the performance of the `RNNCell` implementation. Since I may not have the technical expertise in XLA (sorry for not being very helpful at this stage), I am thinking of a workaround based on `CudnnRNN` bypassing some cell wrappers. Nevertheless, I am happy to hear that XLA:GPU for LSTMs is already on your roadmap.

Will also close the issue unless you consider it is better otherwise.
